ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs

```diff
@@ -764,6 +766,69 @@ runWithTempRegistry m = withRegistry $ \rr -> do
         }
     return a
 
+-- | Embed a self-contained 'WithTempRegistry' computation into a larger one.
+--
+-- The internal 'WithTempRegistry' is effectively passed to
+-- 'runWithTempRegistry'. It therefore must have no dangling resources, for
+-- example. This is the meaning of /self-contained/ above.
+--
+-- The key difference beyond 'runWithTempRegistry' is that the resulting
+-- composite resource is also guaranteed to be registered in the outer
+-- 'WithTempRegistry' computation's registry once the inner registry is closed.
+-- Combined with the following assumption, this establishes the invariant that
+-- all resources are (transitively) in a temporary registry.
+--
+-- As the resource might require some implementation details to be closed, the
+-- function to close it will also be provided by the inner computation.
+--
+-- ASSUMPTION: closing @res@ closes every resource contained in @inner_st@
+--
+-- NOTE: In the current implementation, there will be a brief moment where the
+-- inner registry still contains the inner computation's resources and also the
+-- outer registry simultaneously contains the new composite resource. If an async
+-- is received at that time, then the inner resources will be closed and then
+-- the composite resource will be later closed. This means there's a risk of
+-- /double freeing/, which can be harmless if anticipated.
+runInnerWithTempRegistry
+  :: forall inner_st st m res a. IOLike m
+  => WithTempRegistry inner_st m (a, inner_st, res, res -> m Bool)
+     -- ^ The embedded computation; see ASSUMPTION above
+  -> (st -> res -> Bool)
```

Comment:
In our codebase, we currently don't. But as a part of the logical `ResourceRegistry` api in a vacuum, it makes sense.

In particular, deferring the use of `\_ _ -> True` to the call-site encourage us to write the justification for the degenerate check at the place where we select it (see my other PR comment requesting such Haskell comments). That's better than hiding that degeneracy in this otherwise plain-looking function call.

Reply:
Nick's commits address this concern :+1: